### PR TITLE
Research: pronunciation evidence engine R&D V1

### PR DESCRIPTION
**Pull request by Thunderkill016**
## Goal
Build the first model-independent research kernel for an original AtoEnglish pronunciation-assessment engine. The target is not to copy a vendor score but to preserve acoustic uncertainty, align multiple valid pronunciations, diagnose phone errors articulatorily, and calibrate any future numeric score against human ratings.

## Exception / current blocker
This work is intentionally isolated from the active pilot queue under the `DO_NOT_BUILD.md` exception rule. The blocker occurred in the real development flow twice: (1) the paid/provider pronunciation path failed in practice, and (2) the free browser top-1 phoneme path discarded the posterior evidence needed to distinguish learner errors reliably. Existing simpler solutions are insufficient for the requested free pronunciation-scoring direction because top-1 ASR output cannot expose hidden canonical-phone support or calibrated uncertainty.

The implementation remains reversible and bounded: one draft R&D branch/PR, no merge/deploy, no database/auth/analytics/XP/FSRS/curriculum changes, no raw-audio persistence, and no learner-facing score until human calibration exists. Rollback is closing this draft PR and deleting the branch.

## What is implemented
- frontier research program and license boundaries
- explicit sensor observation / diagnosis / calibration contracts
- English articulatory phone representation and transparent phonological distance prior
- posterior-aware dynamic-programming alignment
- explicit substitution/deletion/insertion evidence
- multiple canonical-pronunciation selection
- rank-only candidate handling without fabricating probabilities
- uncalibrated segmental evidence derivation
- monotonic isotonic calibration (PAVA)
- split-conformal residual intervals
- research metrics: Pearson, Spearman, MAE, RMSE, F1, MCC, FAR/FRR, Brier, ECE, ROC AUC
- raw CTC logits/posterior extraction in the local browser worker
- greedy CTC segment bridge with original posterior mass preserved
- posterior uncertainty and temporal-stability statistics
- local signal-quality and prosody evidence
- deterministic sensor benchmark utilities
- canonical CTC forward-lattice likelihood in log space
- repeated-label blank enforcement and impossible-target validation
- exact path-sum regression test plus blank-dominant canonical-evidence tests

## Scientific boundary
This is **not** a validated pronunciation scorer. No learner-facing 0–100 score is produced. Raw accuracy/completeness signals remain research-only and `calibration: "unvalidated"` until speaker-disjoint human-rated data supports calibration.

The engine explicitly separates:

`observation != diagnosis != calibrated score != learning decision`

## License boundary
- commercially compatible foundations can include SpeechOcean762 (CC BY 4.0), GOPT concepts/code where appropriate (BSD-3-Clause), Kaldi GOP concepts (Apache-2.0 project), and individually reviewed commercial-safe acoustic checkpoints.
- L2-ARCTIC is treated as research-only for AtoEnglish because it is CC BY-NC 4.0.
- IF-MDD and HierTFR are research references only here; their public repositories did not expose a usable repository license during review, so their source is not copied into AtoEnglish.

## Out of scope
- no database or migration changes
- no auth/RLS changes
- no analytics changes
- no XP/FSRS/mastery changes
- no lesson/curriculum changes
- no paid API
- no model training in this PR
- no production deployment
- no merge by agent

## Verification on current head
Head: `d3fe450e58eca3724d1f89fd2378db9dfc2e1492`.

[GitHub Verify run 33948085867](https://github.com/Thunderkill016/AtoEnglish/actions/runs/33948085867) passed on this head:
- lint, type-check, unit tests, and content standards
- fresh migrations, database lint, and pgTAP/RLS tests

Local checks rerun after committing this head:
- focused CTC lattice suite: 16 tests passed
- `npx tsc --noEmit` and `npm run lint`: passed
- `npm run test`: 63 files / 486 tests passed
- `npm run test:content-standard`: 50 tests passed
- `npm run build -- --webpack`: passed (90 pages generated; existing Sentry configuration warnings)

## Completed bounded task: canonical CTC forward lattice
The engine now computes `log P(canonical target | acoustic frames)` by summing all valid CTC paths rather than relying on greedy top-1 output. It preserves evidence for a target phone even when blank wins every frame, enforces the CTC repeated-label blank rule, and rejects impossible/invalid target paths.

## Completed bounded task: CTC forward-backward state posterior V1
`canonicalCtcForwardBackward()` now returns canonical sequence log-likelihood, the extended target, minimum frame count, the complete state posterior matrix, and a state-to-canonical-position/blank mapping. Each canonical position also receives expected occupied frames, expected duration, peak-support frame, and peak posterior. Repeated tokens retain distinct positions.

The backward recurrence uses suffix probabilities excluding the current emission and exactly reverses the forward transition rules. Gamma comes directly from alpha, beta, and sequence likelihood without post-hoc normalization. The existing forward-only API retains rolling-row memory; only the new API retains the lattice.

Tests cover manually enumerated path posteriors, frame mass conservation, blank/competitor dominance, repeated labels, forward likelihood equivalence, duration, nonzero blank IDs, single-frame targets, long-sequence underflow, and explicit invalid/zero-mass failures.

Occupancy is conditional on the supplied target and is not evidence that the pronunciation is correct. All outputs remain research-only; calibration remains unvalidated. This task changes only `ctc-lattice.ts` and `ctc-lattice.test.ts`. The existing barrel export already exposes the new API.

## Next empirical gate
After forward-backward state evidence is verified, benchmark deliberate `θ / s / t` contrasts on identical learner audio. A sensor is discarded if it cannot separate those large contrasts or if it produces high false-alarm rates.

### CODE REVIEW

**Comment by Thunderkill016** [comment]
## Codex task — CTC forward-backward state posterior V1

Work on the existing branch `frontier/pronunciation-engine-rd-v1`. Do **not** merge or deploy. Before changing code, read `AGENTS.md`, `docs/product/PRODUCT_TRUTH.md`, `docs/product/CURRENT_PRIORITY.md`, `docs/product/DO_NOT_BUILD.md`, then inspect the current pronunciation-engine files and tests.

### Problem
The engine now computes canonical CTC forward-lattice likelihood, but it cannot yet localize which frames support each canonical phone. Greedy CTC is insufficient because canonical evidence can exist even when blank or a competing phone wins top-1.

### Intended outcome
Add a numerically stable backward pass and state posterior `gamma(t,s)` for the canonical CTC lattice, then derive research-only per-phone occupancy/timing evidence. This is groundwork for future GOP/MDD experiments; it is **not** a learner-facing score.

### Current verified baseline
- PR #107 is Draft.
- Last green head before this task: `76060e3a8b4c05b8d6a8ec457b4797dece87a672`.
- Verify #477 passed on that head.
- `canonicalCtcLogLikelihood()` already exists in `src/lib/pronunciation-engine/ctc-lattice.ts`.

If the branch has moved, inspect the new diff and adapt; do not reset or overwrite unrelated work.

### Allowed scope
Prefer only:
- `src/lib/pronunciation-engine/ctc-lattice.ts`
- `src/lib/pronunciation-engine/ctc-lattice.test.ts`
- `src/lib/pronunciation-engine/index.ts` only if a new public export is genuinely required

A tiny adjacent type/helper change is allowed only when the implementation cannot remain coherent without it.

### Forbidden scope
Do not change:
- database/migrations/RLS/auth
- analytics/privacy boundaries
- lessons/curriculum
- XP/FSRS/mastery
- dependencies
- browser worker/model loading in this task
- production deployment/config
- learner-facing numeric scoring
- calibration state (`unvalidated` must remain truthful)

### Algorithm contract
Use the existing extended CTC target with blanks and the same legal transition rules as the forward pass.

Implement backward values in log-space. Use a definition equivalent to:
- `alpha[t][s]` includes the emission probability for state `s` at frame `t`.
- `beta[t][s]` represents suffix probability from frames `t+1..T-1`, conditional on being in state `s` at frame `t`, so it excludes the emission at `t`.
- terminal backward states at `T-1`: final blank and final label have log-beta `0`; all other states are `-Infinity`.
- successor transitions from state `s`: stay at `s`, advance to `s+1`, and advance to `s+2` only when the destination is non-blank and is not the same label as state `s` (preserve the repeated-label blank rule).
- `gamma[t][s] = exp(alpha[t][s] + beta[t][s] - logLikelihood)`.

Do not renormalize gamma as a workaround. On valid non-zero-probability targets, the sum of state posteriors for each frame should be approximately `1` from the actual forward-backward math.

### Public research output
Add one coherent API, for example `canonicalCtcForwardBackward(...)` (name may differ if the existing module suggests a better one), returning at minimum:
- `logLikelihood`
- `extendedTarget`
- `minimumRequiredFrames`
- state posterior matrix or an equally lossless representation of `gamma(t,s)`
- mapping from extended-state index to canonical phone position / blank

Also derive a small per-canonical-phone research summary from gamma, at minimum:
- canonical phone/token position
- expected occupied frames (`sum_t gamma[t,labelState]`)
- expected duration in milliseconds using `audioDurationMs / frameCount`
- peak-support frame and peak posterior

Do not call these values `accuracy`, `confidence`, `score`, or `mastery`.

### Correctness requirements
1. Forward-backward `logLikelihood` must agree with the existing forward-only result within floating-point tolerance.
2. For every valid frame, `sum_s gamma[t][s] ≈ 1`.
3. A target phone can have non-zero posterior occupancy even when blank is greedy top-1 on every frame.
4. Repeated canonical labels still require an intervening blank; no illegal skip transition.
5. Impossible / zero-probability targets retain the current explicit failure behavior rather than producing NaN/Infinity output.
6. No probability fabrication and no arbitrary post-hoc normalization.

### Required tests
Add deterministic tests covering at least:
- exact/small lattice where gamma can be checked manually or by strong invariants;
- per-frame gamma sums to 1;
- blank-dominant frames still yield canonical phone occupancy;
- repeated-label target behavior;
- forward-only and forward-backward likelihood equivalence;
- expected-duration calculation from posterior occupancy;
- invalid/zero-mass paths remain rejected.

Prefer mathematical invariants over brittle exact floating-point arrays.

### Required verification
Run on the final committed state:
```bash
npx vitest run src/lib/pronunciation-engine/ctc-lattice.test.ts
npx tsc --noEmit
npm run lint
npm run test
npm run test:content-standard
```
Then rely on the repository PR Verify workflow for fresh migrations / DB lint / pgTAP as usual. Do not claim `npm run build` passed unless you actually run it.

### Manual review questions
- Does beta use exactly the reverse of the legal forward transition graph?
- Does gamma come from the path probability math rather than a second normalization step?
- Can a weak but real canonical phone survive blank-heavy frames?
- Are repeated labels protected from illegal direct transitions?
- Did any research-only value accidentally become learner-facing?

### Rollback
The task must remain isolated in PR #107. If the math cannot be verified cleanly, revert only this bounded forward-backward change and keep the already-green forward lattice.

### Stop condition
Stop after forward-backward + state-posterior + minimal per-phone occupancy/timing evidence is green. Do **not** continue into GOP scoring, competitor diagnosis, human calibration, browser integration, UI, or model training in the same task.

**Comment by Thunderkill016** [comment]
Completed the assigned CTC forward-backward task in commit d3fe450e58eca3724d1f89fd2378db9dfc2e1492.

Only src/lib/pronunciation-engine/ctc-lattice.ts and ctc-lattice.test.ts changed. Beta reverses the legal forward transitions, excludes the current emission, and produces gamma directly without normalization. The API includes position/blank mapping and per-phone expected occupancy, duration, and peak support.

Final committed-state verification: focused 16 tests, full 486 tests, content 50 tests, typecheck, lint, and webpack production build passed. Verify run 33948085867 passed both code and fresh migrations/database lint/pgTAP jobs on the same head:
https://github.com/Thunderkill016/AtoEnglish/actions/runs/33948085867

Manual review: repeated-label skips remain prohibited; blank-heavy canonical support survives; occupancy is explicitly conditional on the target; the API has no learner-facing caller. PR body now reflects the current head and completed task. PR remains Draft. Stopping at the assigned mathematical evidence boundary; no merge or deployment performed.
